src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,329 @@
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // This is a start tag. Keep track of it.
+                        _startTagTracker.Push(new TagBlockTracker(tagBlock));
+                    }
+                }
+
+                while (_startTagTracker.Count > 0)
+                {
+                    // We reached the end of the list and still have unmatched start tags
+                    var startTagTracker = _startTagTracker.Pop();
+                    var startTag = startTagTracker.TagBlock;
+                    var tagChildren = startTagTracker.Children;
+                    BuildMarkupElement(rewrittenChildren, startTag, tagChildren, endTag: null);
+                }
+
+                // We now have finished building our list of rewritten Children.
+                // At this point, We should have a one to one replacement for every child. The replacement can be null.
+                Debug.Assert(children.Count == rewrittenChildren.Count);
+                node = node.ReplaceNodes(children, (original, rewritten) =>
+                {
+                    var originalIndex = children.IndexOf(original);
+                    if (originalIndex != -1)
+                    {
+                        // If this returns null, that node will be removed.
+                        return rewrittenChildren[originalIndex];
+                    }
+
+                    return original;
+                });
+
+                return node;
+            }
+
+            private void BuildMarkupElement(List<SyntaxNode> rewrittenChildren, MarkupTagBlockSyntax startTag, List<RazorSyntaxNode> tagChildren, MarkupTagBlockSyntax endTag)
+            {
+                // We are trying to replace multiple nodes (including the start/end tag) with one rewritten node.
+                // Since we need to have each child node accounted for in our rewritten list,
+                // we'll add "null" in place of them.
+                // The call to SyntaxNode.ReplaceNodes() later will take care removing the nodes whose replacement is null.
+
+                var body = tagChildren.Where(t => t != null).ToList();
+                var rewritten = SyntaxFactory.MarkupElement(startTag, new SyntaxList<RazorSyntaxNode>(body), endTag: endTag);
+                if (startTag != null)
+                {
+                    // If there was a start tag, that is where we want to put our new element.
+                    TrackChild(rewritten, rewrittenChildren);
+                }
+
+                tagChildren.ForEach(c => TrackChild(null, rewrittenChildren));
+                if (endTag != null)
+                {
+                    TrackChild(startTag == null ? rewritten : null, rewrittenChildren);
+                }
+            }
+
+            private void TrackChild(SyntaxNode child, List<SyntaxNode> rewrittenChildren)
```

Comment:
Confusing af

Reply:
Added this to https://github.com/aspnet/Razor/issues/2619.